File: src/Modules/MedicalVisits/ErpClink.Modules.MedicalVisits.Application/Visits/Models/MedicalVisitModels.cs
namespace ErpClink.Modules.MedicalVisits.Application.Visits.Models;

public sealed record StartMedicalVisitRequest(Guid AppointmentId, Guid? QueueEntryId);

public sealed record UpdateClinicalNotesRequest(
    string? ChiefComplaint,
    string? ClinicalNotes,
    string? ExaminationFindings,
    string? DiagnosisNotes,
    string? FollowUpNotes,
    byte[]? RowVersion);

public sealed record CancelMedicalVisitRequest(string? Reason);

public sealed record SearchMedicalVisitsRequest(
    Guid? PatientId,
    Guid? DoctorId,
    Guid? ClinicId,
    Guid? AppointmentId,
    DateOnly? VisitDate,
    DateOnly? DateFrom,
    DateOnly? DateTo,
    string? Status,
    string? VisitNumber,
    int Page = 1,
    int PageSize = 20);

public sealed record PatientHistoryRequest(
    DateOnly? DateFrom,
    DateOnly? DateTo,
    Guid? DoctorId,
    Guid? ClinicId,
    string? Status,
    int Page = 1,
    int PageSize = 20);

public sealed record MedicalVisitDto(
    Guid Id,
    Guid OrganizationId,
    Guid BranchId,
    string VisitNumber,
    Guid PatientId,
    Guid AppointmentId,
    Guid DoctorId,
    Guid ClinicId,
    Guid? QueueEntryId,
    DateOnly VisitDate,
    string Status,
    string? ChiefComplaint,
    string? ClinicalNotes,
    string? ExaminationFindings,
    string? DiagnosisNotes,
    string? FollowUpNotes,
    DateTime? CompletedAtUtc,
    string? CompletedBy,
    DateTime? CancelledAtUtc,
    string? CancelledBy,
    string? CancellationReason,
    DateTime CreatedAtUtc,
    string? CreatedBy,
    DateTime? UpdatedAtUtc,
    string? UpdatedBy,
    byte[] RowVersion);

public sealed record MedicalVisitListItemDto(
    Guid Id,
    string VisitNumber,
    DateOnly VisitDate,
    Guid PatientId,
    Guid DoctorId,
    Guid ClinicId,
    Guid AppointmentId,
    string Status,
    string? ChiefComplaint,
    string? DiagnosisNotes,
    string? FollowUpNotes);

public sealed record PagedMedicalVisitsResult(
    IReadOnlyList<MedicalVisitListItemDto> Items,
    int Page,
    int PageSize,
    int TotalCount);

/// <summary>
/// Derived patient visit context for Patient 360 (not a persisted patient status).
/// Relationship and treatment state are computed from visit history.
/// </summary>
public sealed record PatientVisitContextDto(
    int TotalVisits,
    int CompletedVisits,
    int ActiveVisits,
    bool IsReturningPatient,
    string PatientRelationship,
    string TreatmentState,
    string? CurrentEncounterKind,
    MedicalVisitListItemDto? ActiveVisit,
    MedicalVisitListItemDto? LastVisit,
    bool PreviousVisitHadFollowUpNotes);

public static class PatientVisitContextCodes
{
    public const string RelationshipNew = "New";
    public const string RelationshipReturning = "Returning";

    public const string TreatmentNone = "None";
    public const string TreatmentInProgress = "InTreatment";
    public const string TreatmentCompleted = "TreatmentCompleted";

    public const string EncounterFirstVisit = "FirstVisit";
    public const string EncounterFollowUp = "FollowUp";
    public const string EncounterConsultation = "Consultation";
}

File: frontend/erpclink-web/src/app/features/medical-visits/models/medical-visit.models.ts
export type MedicalVisitStatus = 'Open' | 'InProgress' | 'Completed' | 'Cancelled';

export interface MedicalVisitDto {
  id: string;
  organizationId: string;
  branchId: string;
  visitNumber: string;
  patientId: string;
  appointmentId: string;
  doctorId: string;
  clinicId: string;
  queueEntryId: string | null;
  visitDate: string;
  status: MedicalVisitStatus;
  chiefComplaint: string | null;
  clinicalNotes: string | null;
  examinationFindings: string | null;
  diagnosisNotes: string | null;
  followUpNotes: string | null;
  completedAtUtc: string | null;
  completedBy: string | null;
  cancelledAtUtc: string | null;
  cancelledBy: string | null;
  cancellationReason: string | null;
  createdAtUtc: string;
  createdBy: string | null;
  updatedAtUtc: string | null;
  updatedBy: string | null;
  rowVersion: string;
}

export interface MedicalVisitListItemDto {
  id: string;
  visitNumber: string;
  visitDate: string;
  patientId: string;
  doctorId: string;
  clinicId: string;
  appointmentId: string;
  status: MedicalVisitStatus;
  chiefComplaint: string | null;
  diagnosisNotes: string | null;
  followUpNotes: string | null;
}

export interface PagedMedicalVisitsResult {
  items: MedicalVisitListItemDto[];
  page: number;
  pageSize: number;
  totalCount: number;
}

export interface StartMedicalVisitRequest {
  appointmentId: string;
  queueEntryId?: string | null;
}

export interface UpdateClinicalNotesRequest {
  chiefComplaint?: string | null;
  clinicalNotes?: string | null;
  examinationFindings?: string | null;
  diagnosisNotes?: string | null;
  followUpNotes?: string | null;
  rowVersion?: string | null;
}

export interface CancelMedicalVisitRequest {
  reason?: string | null;
}

export interface SearchMedicalVisitsParams {
  patientId?: string;
  doctorId?: string;
  clinicId?: string;
  appointmentId?: string;
  visitDate?: string;
  dateFrom?: string;
  dateTo?: string;
  status?: MedicalVisitStatus | '';
  visitNumber?: string;
  page?: number;
  pageSize?: number;
}

export interface PatientVisitHistoryParams {
  dateFrom?: string;
  dateTo?: string;
  doctorId?: string;
  clinicId?: string;
  status?: MedicalVisitStatus | '';
  page?: number;
  pageSize?: number;
}

export function isMedicalVisitNotesEditable(status: MedicalVisitStatus): boolean {
  return status === 'Open' || status === 'InProgress';
}

/** Derived Patient 360 visit context (not a persisted patient status). */
export type PatientRelationshipCode = 'New' | 'Returning';
export type PatientTreatmentStateCode = 'None' | 'InTreatment' | 'TreatmentCompleted';
export type CurrentEncounterKindCode = 'FirstVisit' | 'FollowUp' | 'Consultation';

export interface PatientVisitContextDto {
  totalVisits: number;
  completedVisits: number;
  activeVisits: number;
  isReturningPatient: boolean;
  patientRelationship: PatientRelationshipCode;
  treatmentState: PatientTreatmentStateCode;
  currentEncounterKind: CurrentEncounterKindCode | null;
  activeVisit: MedicalVisitListItemDto | null;
  lastVisit: MedicalVisitListItemDto | null;
  previousVisitHadFollowUpNotes: boolean;
}

export const PATIENT_RELATIONSHIP_LABELS: Record<PatientRelationshipCode, string> = {
  New: 'مريض جديد',
  Returning: 'مريض عائد'
};

export const PATIENT_TREATMENT_STATE_LABELS: Record<PatientTreatmentStateCode, string> = {
  None: 'لا توجد زيارات بعد',
  InTreatment: 'قيد العلاج',
  TreatmentCompleted: 'اكتمل العلاج'
};

export const CURRENT_ENCOUNTER_KIND_LABELS: Record<CurrentEncounterKindCode, string> = {
  FirstVisit: 'زيارة أولى',
  FollowUp: 'متابعة',
  Consultation: 'استشارة'
};
